src/Search.php

```diff
@@ -0,0 +1,154 @@
+use \Exception;
+
+/*
+ * ahocorasick - fast string searching in php
+ *
+ * Copyright (c) 2017-2018 Lewis Van Winkle
+ *
+ * http://CodePlea.com
+ *
+ * This software is provided 'as-is', without any express or implied
+ * warranty. In no event will the authors be held liable for any damages
+ * arising from the use of this software.
+ *
+ * Permission is granted to anyone to use this software for any purpose,
+ * including commercial applications, and to alter it and redistribute it
+ * freely, subject to the following restrictions:
+ *
+ * 1. The origin of this software must not be misrepresented; you must not
+ *    claim that you wrote the original software. If you use this software
+ *    in a product, an acknowledgement in the product documentation would be
+ *    appreciated but is not required.
+ * 2. Altered source versions must be plainly marked as such, and must not be
+ *    misrepresented as being the original software.
+ * 3. This notice may not be removed or altered from any source distribution.
+ *
+ */
+
+
+
+class Search
+{
+
+    private $nodes = [ [] ];
+
+    private $final = 0;
+
+    /**
+     * Add a keyword to search for.
+     *
+     * @param string $needle
+     * @throws Exception
+     */
+    public function addNeedle(string $needle)
+    {
+        if ($this->final) {
+            throw new Exception('Cannot add word to finalized ahocorasick.');
+        }
+
+        $nodes = &$this->nodes;
+        $n = 0;
+
+        $needleLength = strlen($needle);
+
+        for ($i = 0; $i < $needleLength; ++$i) {
+            $c = $needle[$i];
+
+            if (!isset($nodes[$n][$c])) {
+                $nodes[$n][$c] = count($nodes);
+                $nodes[] = [];
+            }
+            $n = $nodes[$n][$c];
+        }
+
+        $nodes[$n][0][] = $needle;
+    }
+
+    /**
+     * Create the structure needed to search text for the given keywords.
+     * Once you call this, you cannot add additional keywords via addNeedle().
+     */
+    public function finalize()
+    {
+        $nodes = &$this->nodes;
+        $queue = [];
+
+        foreach ($nodes[0] as $j => $_) {
+            $nodes[$nodes[0][$j]][1] = 0;
+            $queue[] = $nodes[0][$j];
+        }
+
+        while (count($queue)) {
+            $r = $queue[0];
+            $queue = array_slice($queue, 1);
+
+            foreach ($nodes[$r] as $j => $_) {
+                if ($j === 0 || $j === 1) {
+                    continue;
+                }
+                $v = $nodes[$r][1];
+                $u = $nodes[$r][$j];
+                while ($v > 0 && !isset($nodes[$v][$j])) {
+                    $v = $nodes[$v][1];
+                }
+                $nodes[$u][1] = $nodes[$v][$j] ?? $v;
+                if (isset($nodes[$nodes[$u][1]][0])) {
+                    if (!isset($nodes[$u][0])) {
+                        $nodes[$u][0] = [];
+                    }
+                    $nodes[$u][0] = array_merge($nodes[$u][0], $nodes[$nodes[$u][1]][0]);
```

Comment:
Static analysis identified this as another [possible optimization](https://github.com/kalessil/phpinspectionsea/blob/master/docs/performance.md#slow-array-function-used-in-loop). I haven't touched it yet where I don't have a full grasp of the code at this time and how it would relate to moving the merge process out of the loop.